elf_reader.go

```diff
@@ -247,6 +223,47 @@ func newElfSection(section *elf.Section, kind elfSectionKind) *elfSection {
 	}
 }
 
+// assignSymbols takes a list of symbols and assigns them to their
+// respective sections, indexed by name.
+func (ec *elfCode) assignSymbols(symbols []elf.Symbol) {
+	for _, symbol := range symbols {
+		symType := elf.ST_TYPE(symbol.Info)
+		symSection := ec.sections[symbol.Section]
+		if symSection == nil {
+			continue
+		}
+
+		// Anonymous symbols only occur in debug sections which we don't process
+		// relocations for. Anonymous symbols are not referenced from other sections.
+		if symbol.Name == "" {
+			continue
+		}
+
+		// Older versions of LLVM don't tag symbols correctly, so keep
+		// all NOTYPE ones.
+		switch symSection.kind {
+		case mapSection, btfMapSection, dataSection:
+			if symType != elf.STT_NOTYPE && symType != elf.STT_OBJECT {
+				continue
+			}
+		case programSection:
+			if symType != elf.STT_NOTYPE && symType != elf.STT_FUNC {
+				continue
+			}
+			// Local Basic Blocks are LLVM-internal symbols used as jump
+			// targets within sections and can be ignored.
+			if strings.HasPrefix(symbol.Name, "LBB") {
```

Comment:
The issue is with clang7 that seems to emit relocations as NOTYPE/LOCAL:
```
llvm-readelf -a testdata/loader-clang-7-el.elf | grep NOTYPE | grep LOCAL
     0: 0000000000000000     0 NOTYPE  LOCAL  DEFAULT   UND
     1: 0000000000000000     0 NOTYPE  LOCAL  DEFAULT     2 static_fn
     
llvm-readelf -a testdata/loader-clang-13-el.elf | grep NOTYPE | grep LOCAL
     0: 0000000000000000     0 NOTYPE  LOCAL  DEFAULT   UND
    12: 00000000000000e8     0 NOTYPE  LOCAL  DEFAULT     7 LBB7_5
```

I just noticed LBB symbols carry some kind of value, and skipping NOTYPE && LOCAL && value != 0 seems to work.

Reply:
Dropping static isn't as straight forward I'd say, since it says something about the visibility of a function. Do you refer to not support static on global variables?

Maybe the thing to do is to deprecate clang-7 support instead?

Reply:
https://releases.llvm.org/ says clang 7.1.0 was released 10 May 2019 probably too soon to get rid of it. NOTYPE && LOCAL && string match should be good enough.

Reply:
Okay, will do. Unfortunately, the string match will actually fire in the majority of cases in the Cilium datapath, since LBB_ symbols vastly outnumber any other kind of symbol. :disappointed: